Migrate System.Numerics.Vectors

carlossanlop opened the pull request:
Still in draft because there are build failures and need guidance on how to get the tempates to work.

The [System.Numerics.Vectors nuget package](https://www.nuget.org/packages/System.Numerics.Vectors) has this information:

- Has public types only in .NET Framework 4.6.2.
- Supports .NET Standard 1.0 and 2.0.
- Has an empty placeholder `_._` in the `lib/netcoreapp2.0` folder.
- Has no ref assemblies.
- Has no PNSE.

When consuming this nuget package in a console app, then printing `typeof( System.Numerics.Matrix4x4).Assembly.CodeBase` I confirmed that:

- When targeting .NET Framework 4.6.2, this assembly is used: `C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Numerics/v4.0_4.0.0.0__b77a5c561934e089/System.Numerics.dll`
- When targeting .NET 6.0, this assembly is used: `C:/Program Files/dotnet/shared/Microsoft.NETCore.App/6.0.27/System.Private.CoreLib.dll`
- BUT... When targeting netstandard2.0, this assembly is used: `bin\Debug\netstandard2.0\System.Numerics.Vectors.dll`

carlossanlop commented:
A bunch of the build failures are happening due to the conflict of the System.Numerics types (like the type `Matrix4x4`) which are getting imported in net462 by default (the one that is in-box: `C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.2\System.Numerics.dll`).

I tried addingthis to the src csproj without success:

```xml
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
    <Reference Remove="System.Numerics" />
  </ItemGroup>
```

Then I realized I might have misunderstood which tfm is consuming the nuget package. According to the PR description, only netstandard2.0 is using the nuget package. Both net6.0 and net462 use the in-box assembly.

This is the first time I see a migrated project do this. Is this a package we still want to migrate, @ericstj @ViktorHofer ?

ViktorHofer commented on src/System.Numerics.Vectors/tests/System.Numerics.Vectors.Tests.csproj:
```diff
@@ -0,0 +1,58 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(NetFrameworkMinimum)</TargetFrameworks>
```
```suggestion
    <TargetFramework>$(NetFrameworkMinimum)</TargetFramework>
```

ericstj commented:
Double check your conditions on your compile sources.  This assembly does carry some implementation types for .NETFramework.  See https://github.com/dotnet/corefx/blob/06a86ddfefa9042b22f2bcefc9e32088e077df78/src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj#L31-L81

It's a partial facade - some types are forwarded, some types are defined.

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -0,0 +1,74 @@
+    </Content>
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(IsPlaceholderTargetFramework)' != 'true'">
```
Looks to me like *some* of these were also excluded on NetFx before
https://github.com/dotnet/corefx/blob/06a86ddfefa9042b22f2bcefc9e32088e077df78/src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj#L70-L81

carlossanlop commented on src/System.Numerics.Vectors/tests/Vector2Tests.cs:
```diff
@@ -0,0 +1,1199 @@
+            Assert.Throws<ArgumentOutOfRangeException>(() => v1.CopyTo(a, -1));
+            Assert.Throws<ArgumentOutOfRangeException>(() => v1.CopyTo(a, a.Length));
+            
+            if (!PlatformDetection.IsNetNative)
```
This PlatformDetection property does not currently exist. Should I add it? The value is:

```cs
        public static bool IsNetNative => RuntimeInformation.FrameworkDescription.StartsWith(".NET Native", StringComparison.OrdinalIgnoreCase);
```

But I am unsure if we need it.

For now, I'm going to remove it, as it is affecting the ability to build.

ericstj replied:
Depends on what's there and how much of the test codebase needs it.  We're getting close to the end now, so maybe folks didn't need it?

carlossanlop replied:
Well, I only found the one use. And haven't seen it in any of my other PRs (yet).

ericstj replied:
Oh, I see - it's only a single property.  Delete the code :)

carlossanlop replied:
Deleted

carlossanlop commented on src/System.Numerics.Vectors/tests/Matrix4x4Tests.cs:
```diff
@@ -0,0 +1,2549 @@
+            Assert.Equal(new IntPtr(basePtr + 15), new IntPtr(&mat.M44));
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public void PerspectiveFarPlaneAtInfinityTest()
+        {
+            var nearPlaneDistance = 0.125f;
+            var m = Matrix4x4.CreatePerspective(1.0f, 1.0f, nearPlaneDistance, float.PositiveInfinity);
+            Assert.Equal(-1.0f, m.M33);
+            Assert.Equal(-nearPlaneDistance, m.M43);
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public void PerspectiveFieldOfViewFarPlaneAtInfinityTest()
+        {
+            var nearPlaneDistance = 0.125f;
+            var m = Matrix4x4.CreatePerspectiveFieldOfView(MathHelper.ToRadians(60.0f), 1.5f, nearPlaneDistance, float.PositiveInfinity);
+            Assert.Equal(-1.0f, m.M33);
+            Assert.Equal(-nearPlaneDistance, m.M43);
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public void PerspectiveOffCenterFarPlaneAtInfinityTest()
+        {
+            var nearPlaneDistance = 0.125f;
+            var m = Matrix4x4.CreatePerspectiveOffCenter(0.0f, 0.0f, 1.0f, 1.0f, nearPlaneDistance, float.PositiveInfinity);
+            Assert.Equal(-1.0f, m.M33);
+            Assert.Equal(-nearPlaneDistance, m.M43);
+        }
```
The SkipOnTargetFramework logic is not working. These 3 tests are getting executed in .NET Framework and failing. Since the src csproj has a netcoreapp2.0 tfm just to cover the `_._` case, I don't think I need to add that tfm to the tests csproj. We only need to test `$(NetFrameworkMinimum)`, correct? So I can just delete these 3 tests:

```
[xUnit.net 00:00:01.68]     System.Numerics.Tests.Matrix4x4Tests.PerspectiveFieldOfViewFarPlaneAtInfinityTest [FAIL]
  Failed System.Numerics.Tests.Matrix4x4Tests.PerspectiveFieldOfViewFarPlaneAtInfinityTest [10 ms]
  Error Message:
   Assert.Equal() Failure
Expected: -1
Actual:   NaN
  Stack Trace:
     at System.Numerics.Tests.Matrix4x4Tests.PerspectiveFieldOfViewFarPlaneAtInfinityTest() in C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\tests\Matrix4x4Tests.cs:line 2535
[xUnit.net 00:00:01.72]     System.Numerics.Tests.Matrix4x4Tests.PerspectiveOffCenterFarPlaneAtInfinityTest [FAIL]
  Failed System.Numerics.Tests.Matrix4x4Tests.PerspectiveOffCenterFarPlaneAtInfinityTest [1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: -1
Actual:   NaN
  Stack Trace:
     at System.Numerics.Tests.Matrix4x4Tests.PerspectiveOffCenterFarPlaneAtInfinityTest() in C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\tests\Matrix4x4Tests.cs:line 2545
[xUnit.net 00:00:01.73]     System.Numerics.Tests.Matrix4x4Tests.PerspectiveFarPlaneAtInfinityTest [FAIL]
  Failed System.Numerics.Tests.Matrix4x4Tests.PerspectiveFarPlaneAtInfinityTest [1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: -1
Actual:   NaN
  Stack Trace:
     at System.Numerics.Tests.Matrix4x4Tests.PerspectiveFarPlaneAtInfinityTest() in C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\tests\Matrix4x4Tests.cs:line 2525
```

ericstj replied:
We should try to make SkipOnTargetFramework work, can you make sure to file an issue for that?

carlossanlop replied:
Done https://github.com/dotnet/maintenance-packages/issues/64

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -0,0 +1,87 @@
+    <Compile Include="System\Runtime\CompilerServices\IntrinsicAttribute.cs" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' != '$(NetFrameworkMinimum)'">
```
I think you should also exclude on `IsPlaceholderTargetFramework` - since we don't compile anything there.

carlossanlop replied:
Fixed

ericstj commented on src/System.Numerics.Vectors/src/CompatibilitySuppressions.xml:
```diff
@@ -0,0 +1,60 @@
+    <DiagnosticId>CP0001</DiagnosticId>
+    <Target>T:System.Numerics.Matrix3x2</Target>
+    <Left>lib/netstandard2.0/System.Numerics.Vectors.dll</Left>
+    <Right>lib/net462/System.Numerics.Vectors.dll</Right>
```
You need to add typeforwards for all these types.  Previously the partial facade infrastructure was doing that, but we're not using that anymore, so you'll need a netframework specific source-file that forwards each of these types.

carlossanlop replied:
Is this something that I can generate with a command, or do I need to do this manually?

ericstj replied:
GenAPI and GenFacades can both produce it.  Probably taking us longer to talk about it than just write the 7 lines...  Here's an intermediate file from dotnet/runtime "C:\src\dotnet\runtime\artifacts\obj\System.Numerics.Vectors\Debug\net9.0\System.Numerics.Vectors.Forwards.cs"
```c#
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Matrix3x2))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Matrix4x4))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Plane))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Quaternion))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Vector2))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Vector3))]
[assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Numerics.Vector4))]
```

carlossanlop replied:
Added.

ericstj commented on src/System.Numerics.Vectors/src/CompatibilitySuppressions.xml:
```diff
@@ -0,0 +1,60 @@
+    <Right>lib/net462/System.Numerics.Vectors.dll</Right>
+  </Suppression>
+  <Suppression>
+    <DiagnosticId>CP0003</DiagnosticId>
```
https://learn.microsoft.com/en-us/dotnet/fundamentals/apicompat/diagnostic-ids

| CP0003 | Some part of the assembly identity (name, public key token, culture, retargetable attribute, or version) does not match on both sides of the comparison. | Update the assembly identity so that both sides match. |

Seems like one we want to fix.

carlossanlop replied:
Fixed by setting the key to Microsoft.

ericstj reviewed (changes requested):
Please make the fixes to assembly name and surface area and ensure APICompat is clean.

carlossanlop commented:
> Please make the fixes to assembly name and type forwards and ensure APICompat is clean.

Done. Pending squashing the commits once I get a sign-off.

- 1084 tests are being executed and passing locally:
```
dotnet test .\src\System.Numerics.Vectors\tests\System.Numerics.Vectors.Tests.csproj
...
Passed!  - Failed:     0, Passed:  1084, Skipped:     0, Total:  1084, Duration: 1 s - System.Numerics.Vectors.Tests.dll (net462)
```

- I deleted the suppressions file, fixed CP0003 (the key had to be set to Microsoft), and a new dotnet pack with IsPackable=true did not show CP0003 again. I added the other ones (CP0001) to the new suppressions file.

```
dotnet pack .\src\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj /p:ApiCompatGenerateSuppressionFile=true
MSBuild version 17.9.4+90725d08d for .NET
  Determining projects to restore...
  Restored C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj (in 252 ms).
  System.Numerics.Vectors -> C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Numerics.Vectors\Release\net462\System.Numerics.Vectors.dll
  System.Numerics.Vectors -> C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Numerics.Vectors\Release\netstandard2.0\System.Numerics.Vectors.dll
  The package System.Numerics.Vectors.4.6.0-dev.24129.1 is missing a readme. Go to https://aka.ms/nuget/authoring-best-practices/readme to learn why package readmes are important.
  Successfully created package 'C:\Users\calope\source\repos\maintenance-packages\artifacts\packages\Release\Shipping\System.Numerics.Vectors.4.6.0-dev.24129.1.nupkg'.
  Successfully wrote compatibility suppressions to 'C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\src\CompatibilitySuppressions.xml'.
```

carlossanlop commented:
Just to be clear, if I delete the suppresions file and run dotnet pack again with IsPackable=true, this is what I get:

```
dotnet pack .\src\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj
MSBuild version 17.9.4+90725d08d for .NET
  Determining projects to restore...
  Restored C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj (in 339 ms).
  System.Numerics.Vectors -> C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Numerics.Vectors\Release\net462\System.Numerics.Vectors.dll
  System.Numerics.Vectors -> C:\Users\calope\source\repos\maintenance-packages\artifacts\bin\System.Numerics.Vectors\Release\netstandard2.0\System.Numerics.Vectors.dll
  The package System.Numerics.Vectors.4.6.0-dev.24129.1 is missing a readme. Go to https://aka.ms/nuget/authoring-best-practices/readme to learn why package readmes are important.
  Successfully created package 'C:\Users\calope\source\repos\maintenance-packages\artifacts\packages\Release\Shipping\System.Numerics.Vectors.4.6.0-dev.24129.1.nupkg'.
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API compatibility errors between 'ref/netsta
ndard2.0/System.Numerics.Vectors.dll' (C:\.tools\.nuget\packages\system.numerics.vectors\4.5.0\system.numerics.vectors.4.5.0.nupkg) and 'lib/netstandard2.0/System.Numerics.Vectors.dll' (C:\Users\cal
ope\source\repos\maintenance-packages\artifacts\packages\Release\Shipping\System.Numerics.Vectors.4.6.0-dev.24129.1.nupkg): [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vec
tors\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Matrix3x2' exist
s on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vector
s\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Matrix4x4' exist
s on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vector
s\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Plane' exists on
 [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\sr
c\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Quaternion' exis
ts on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vecto
rs\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector2' exists
on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector3' exists
on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector4' exists
on [Baseline] ref/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API compatibility errors between 'lib/netsta
ndard2.0/System.Numerics.Vectors.dll' (C:\.tools\.nuget\packages\system.numerics.vectors\4.5.0\system.numerics.vectors.4.5.0.nupkg) and 'lib/netstandard2.0/System.Numerics.Vectors.dll' (C:\Users\cal
ope\source\repos\maintenance-packages\artifacts\packages\Release\Shipping\System.Numerics.Vectors.4.6.0-dev.24129.1.nupkg): [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vec
tors\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Matrix3x2' exist
s on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vector
s\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Matrix4x4' exist
s on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vector
s\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Plane' exists on
 [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\sr
c\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Quaternion' exis
ts on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vecto
rs\src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector2' exists
on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector3' exists
on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0001: Type 'System.Numerics.Vector4' exists
on [Baseline] lib/netstandard2.0/System.Numerics.Vectors.dll but not on lib/netstandard2.0/System.Numerics.Vectors.dll [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\
src\System.Numerics.Vectors.csproj]
C:\.tools\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-preview.2.24102.10\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API breaking changes found. If those are int
entional, the APICompat suppression file can be updated by rebuilding with '/p:ApiCompatGenerateSuppressionFile=true' [C:\Users\calope\source\repos\maintenance-packages\src\System.Numerics.Vectors\s
rc\System.Numerics.Vectors.csproj]
```

carlossanlop commented:
@ericstj CP0001 - There's a type that exists on the left side of netstandard2.0 but does not exist on the right side. But... it's comparing a `ref` on the left with a `lib` on the right. I don't have a `ref` on the right, but that's on purpose. It's my understanding that they can all be ignored.

```
API compatibility errors between 'ref/netstandard2.0/System.Numerics.Vectors.dll' (C:\.tools\.nuget\packages\system.numerics.vectors\4.5.0\system.numerics.vectors.4.5.0.nupkg) and 'lib/netstandard2.0/System.Numerics.Vectors.dll'
```

carlossanlop commented:
@ericstj That's the top half of the errors. The bottom half of the errors says there's a difference in netstandard2.0 between the left side of `lib` and the right side of `lib`. That one I'm not sure about, it's the same APIs. Let me inspect the generated DLL.

carlossanlop commented:
Ok I think I got it: The condition to include the Compile items of the types that are missing in netstandard2.0 is wrong. I should explicitly import those for netstandard2.0 only, instead of conditioning netstandard2.0 indirectly.

carlossanlop commented:
I found the problem by inspecting `C:\Users\calope\source\repos\maintenance-packages\artifacts\obj\System.Numerics.Vectors\Release\netstandard2.0\System.Numerics.Vectors.dll` with ILSpy and discovering that all those types were not defined there.

After changing the csproj condition, I refreshed ILSpy and I can see the types now. Running dotnet pack with IsPackable=true shows no errors.

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -59,7 +59,7 @@
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFramework)' != '$(NetFrameworkMinimum)' and '$(IsPlaceholderTargetFramework)' == 'true'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
```
This is fine, it would have also been OK to change to `'$(TargetFramework)' != '$(NetFrameworkMinimum)' and '$(IsPlaceholderTargetFramework)' != 'true'`

carlossanlop replied:
Good catch. I mistakenly both were != originally. Since both conditions mean the same, I think I prefer to leave it as it is now since it's more explicit.

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -0,0 +1,92 @@
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(IsPlaceholderTargetFramework)' == 'true'">
+    <ReferenceFromRuntime Include="System.Private.CoreLib" />
```
This does nothing and should be removed.  It was only important when we were actually compiling something for netcoreapp.

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -0,0 +1,92 @@
+    <IsPackable>false</IsPackable>
+    <VersionPrefix>4.5.0</VersionPrefix>
+    <VersionPrefix Condition="'$(IsPackable)' == 'true'">4.6.0</VersionPrefix>
+    <AssemblyVersion>4.1.4.0</AssemblyVersion>
+    <!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
```
```suggestion
    <AssemblyVersion>4.1.4.0</AssemblyVersion> <!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
```
We put this on the same line elsewhere: https://github.com/dotnet/maintenance-packages/blob/df0d9ffb05a586fdd6362c9ad5aff3135be9a64f/src/System.Memory/src/System.Memory.csproj#L11

Putting it on a new line makes it look like it applies to the following element which is not what we want.

carlossanlop replied:
Yes, oops. My bad.

ericstj commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -0,0 +1,92 @@
+    <AssemblyVersion Condition="'$(IsPackable)' == 'true' and $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.1.5.0</AssemblyVersion>
+    <PackageValidationBaselineVersion>4.5.0</PackageValidationBaselineVersion>
+    <IsPlaceholderTargetFramework Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPlaceholderTargetFramework>
+    <IsPartialFacadeAssembly Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp' or $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">true</IsPartialFacadeAssembly>
```
I think this property does nothing in this repo and can be removed.  Instead you manually defined the type forwards.

ericstj commented on src/System.Numerics.Vectors/tests/System.Numerics.Vectors.Tests.csproj:
```diff
@@ -0,0 +1,65 @@
+  <PropertyGroup>
+    <TargetFramework>$(NetFrameworkMinimum)</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <EnableDefaultItems>false</EnableDefaultItems>
```
Not blocking - but why did we disable default items?  Looks to me like they could work (maybe with Update item for GenericVectorTests.cs to add the metadata).

carlossanlop replied:
It was getting complicated to condition the item groups differently.
I can address it in a separate PR: https://github.com/dotnet/maintenance-packages/issues/66

ericstj reviewed (approved):
LGTM with some minor feedback
